ARrehab/ARrehab/TileGrid.swift

```diff
@@ -0,0 +1,163 @@
+    var zLength: Float = 0.0
+        
+    init(surfaceAnchor: ARPlaneAnchor) {
+        self.surfaceAnchor = surfaceAnchor
+        
+        self.gridEntity = AnchorEntity(anchor: surfaceAnchor)
+        self.gridEntity.transform.translation = surfaceAnchor.center
+        
+        self.generatePossibleTiles()
+    }
+    
+    /*
+     Uses the estimated x and z extents of the surface plane to generate an appropriate amount of tiles with adjusted translations (so that tiles don't extend past edges of plane)
+    */
+    func generatePossibleTiles() {
+        let xExtent = self.surfaceAnchor.extent.x
+        let zExtent = self.surfaceAnchor.extent.z
+        
+        let xSize = Tile.TILE_SIZE.x
+        let zSize = Tile.TILE_SIZE.z
+        
+        var currentX = xExtent/2
+        var currentZ = zExtent/2
+        
+        //Generate the tiles
+        while abs(currentX) <= xExtent/2 {
+            while abs(currentZ) <= zExtent/2 {
+                self.generateOneTile(currentX: currentX, currentZ: currentZ)
+                currentZ -= zSize
+            }
+            currentZ = zExtent/2
+            currentX -= xSize
+        }
+        
+        self.xLength = getMaxX().0.x
+        self.zLength = getMaxZ().0.z
+    }
+    
+    /*
+     Generates a tile with the given coordinates
+     Adds the tile to the possibleTiles dict. and to the gridEntity
+     */
+    func generateOneTile(currentX: Float, currentZ: Float) {
+        let newTile = Tile(name: String(format: "Tile (%f,%f)", currentX, currentZ), x: currentX, z: currentZ, materials: [TileGrid.gridMaterial], adjustTranslation: true)
+        self.possibleTiles[newTile.coords] = newTile
+        self.gridEntity.addChild(newTile)
+    }
+    
+    /*
+     Updates the tile grid if the plane has expanded sufficiently
+     */
+    func updateBoard(updatedAnc: ARPlaneAnchor) {
+        //Update the stored anchor
+        self.surfaceAnchor = updatedAnc
+        
+        /*
+         Check if the new plane is long enough to fit more tiles
+         If so, remove all current tiles and regenerate grid to fit
+         Lastly, adjust translation so grid is re-centered correctly
+         */
+        if (self.surfaceAnchor.extent.x > self.xLength * 2 + Tile.TILE_SIZE.x || self.surfaceAnchor.extent.z > self.zLength * 2 + Tile.TILE_SIZE.z) {
+            
+            self.gridEntity.children.removeAll()
+            self.generatePossibleTiles()
+            
+            self.gridEntity.transform.translation = self.surfaceAnchor.center
+        }
+    }
+    
+    //TODO -- IMPLEMENT DIRECTIONAL KNOWLEDGE
+    func updateBoardOutline(centerTile: Tile) {
+        
+        //Clear current outline
+        self.clearOutline()
+        
+        //Find the tile that would form the upper-right corner of the board, with the given centerTile
+        let cornerCoords = Tile.Coordinates(x: centerTile.coords.x + (1 * Tile.TILE_SIZE.x), z: centerTile.coords.z + (2 * Tile.TILE_SIZE.z))
```

Comment:
What does this do? Where does it take into account the size of the board?

Reply:
First step is to find the tile that is "one in the x-direction" and "two in the z-direction" tiles away from the passed in center tile (which is passed in based on player collision). Note: These are hard coded for the current 3x5 dimensions, and it would be good to make these variables (part of the reason why Tile.DIMENSIONS exists).
After finding the correct coordinates for the corner tile, the next step is to find a tile that is part of the grid which is at this position. That is what the closure assigned to cornerTile does. It iterates through the list of possible tiles and checks their coordinates against the corner coordinates we determined would be necessary for the given center tile. This closure will only return a tile if one exists that has coordinates (approximately) equal to our corner coordinates. If no tile exists (aka the player is too near the edges of the grid) the board won't generate (this is where we take the size of the board into account).
Note: I use an "isApproxEqual" function because I noticed while debugging that the coordinates were sometimes off by ~.05 for some reason. This was the remedy to that (checks for tiles who's coords are within .1 of the requested tile). However, now that I am thinking about it again, I don't understand why this should be necessary... I'll look more into it.